File: subsystems/CANRollerSubsystem.py
# Copyright (c) FIRST and other WPILib contributors.
# Open Source Software; you can modify and/or share it under the terms of
# the WPILib BSD license file in the root directory of this project.F

import commands2
import constants
import rev


class CANRollerSubsystem(commands2.Subsystem):
    def __init__(self) -> None:
        super().__init__()

        self.rollerMotor = rev.SparkMax(
            constants.ROLLER_MOTOR_ID, rev.SparkBase.MotorType.kBrushed
        )

        # set can timeouts. This program only sets parameters on startup and
        # doesn't get any parameters so a long timeout is acceptable. Programs
        # which set or get parameters during runtime likely want a timeout
        # closer or equal to the default.
        self.rollerMotor.setCANTimeout(constants.CAN_TIMEOUT)

        self.sparkConfig = rev.SparkMaxConfig()
        # enable voltage compensation. This makes the performance more consistent
        # at different levels of battery charge at the cost of some peak performance
        # with a fully charged battery. Because the roller never really needs the full
        # speed, the compensation value is set very conservatively
        self.sparkConfig.voltageCompensation(constants.ROLLER_MOTOR_VOLTAGE_COMP)
        self.sparkConfig.smartCurrentLimit(constants.ROLLER_MOTOR_CURRENT_LIMIT)
        self.rollerMotor.configure(
            self.sparkConfig,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )

    # function to run the roller with joystick inputs
    def runRoller(self, forward: float, reverse: float) -> None:
        self.rollerMotor.set(forward - reverse)

File: src/main/java/frc/robot/subsystems/CANRollerSubsystem.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants.RollerConstants;
import com.ctre.phoenix.motorcontrol.can.TalonSRX;
import com.ctre.phoenix.motorcontrol.TalonSRXControlMode;

/** Class to run the rollers over CAN */
public class CANRollerSubsystem extends SubsystemBase {
  private final TalonSRX rollerMotor;

  public CANRollerSubsystem() {
    // Set up the roller motor as a brushed motor
    rollerMotor = new TalonSRX(RollerConstants.ROLLER_MOTOR_ID);
    rollerMotor.configContinuousCurrentLimit(RollerConstants.ROLLER_MOTOR_CURRENT_LIMIT);
    rollerMotor.configVoltageCompSaturation(RollerConstants.ROLLER_MOTOR_VOLTAGE_COMP);
  }

  @Override
  public void periodic() {
  }

  /** This is a method that makes the roller spin */
  public void runRoller(double forward, double reverse) {
    rollerMotor.set(TalonSRXControlMode.PercentOutput, forward - reverse);
  }
}
